omp/distributed/partition_kernels.cpp

```diff
@@ -0,0 +1,102 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/distributed/partition_kernels.hpp"
+
+
+#include <omp.h>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+namespace partition {
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+void build_starting_indices(std::shared_ptr<const DefaultExecutor> exec,
+                            const GlobalIndexType* range_offsets,
+                            const int* range_parts, size_type num_ranges,
+                            int num_parts, int& num_empty_parts,
+                            LocalIndexType* ranks, LocalIndexType* sizes)
+{
+    std::fill_n(sizes, num_parts, 0);
+    auto num_threads = static_cast<size_type>(omp_get_max_threads());
+    auto size_per_thread = (num_ranges + num_threads - 1) / num_threads;
+    vector<LocalIndexType> local_sizes(num_parts * num_threads, 0, {exec});
+#pragma omp parallel
+    {
+        auto thread_id = static_cast<size_type>(omp_get_thread_num());
+        auto thread_begin = size_per_thread * thread_id;
+        auto thread_end = std::min(num_ranges, thread_begin + size_per_thread);
+        auto base = num_parts * thread_id;
+        // local exclusive prefix sum
+        for (auto range = thread_begin; range < thread_end; range++) {
+            auto begin = range_offsets[range];
+            auto end = range_offsets[range + 1];
+            auto part = range_parts[range];
+            ranks[range] = local_sizes[part + base];
+            local_sizes[part + base] += end - begin;
+        }
+#pragma omp barrier
+        // exclusive prefix sum over local sizes
+#pragma omp for reduction(+ : num_empty_parts)
+        for (comm_index_type part = 0; part < num_parts; ++part) {
+            LocalIndexType size{};
+            for (size_type thread = 0; thread < num_threads; ++thread) {
+                auto idx = num_parts * thread + part;
+                auto local_size = local_sizes[idx];
+                local_sizes[idx] = size;
+                size += local_size;
+            }
+            sizes[part] = size;
+            num_empty_parts += size == 0 ? 1 : 0;
+        }
+        // add global baselines to local ranks
```

Comment:
I don't think that is necessary, since we don't use the `nowait` clause

Reply:
Might make sense to double-check that, I've been bitten by such an assumption before (which is also why we have a barrier above)

Reply:
@MarcelKoch Could you point me where I can get the `omp for` has sync in the end?
the previous for is normal for in each thread not `#omp for` so it needs the sync, I think. Or the `#omp for` also has a sync in the beginning

Reply:
I've looked at the 4.5 standard specification https://www.openmp.org/wp-content/uploads/openmp-4.5.pdf, and I interpret the section on p.205 l.17-22 on the `reduction` clause as if it will synchronize the result at the end of the reduction.
But perhaps that is not true if we rely on an older standard.

Reply:
from 3.1, https://www.openmp.org/wp-content/uploads/OpenMP3.1.pdf
p105 l.13 - 19 gives the similar behavior